WIP: Replay (#6)
temporalio/temporal-jumpstart-java: docs/curriculum/Versions.md

   3. Crashes
2. The **previous** Version
   1. Picks up the Workflow Task to resume the Execution
   2. The Task fails with a `NonDeterminismException` (NDE)

Your Workflow will not be corrupted, but you could see an impact on performance and Task
failures appear in the history for this Execution.

One mitigation of this would be to always enforce Versioning for the entire Workflow Definition before any deployment, as 
proposed [here](#user-content-global-workflow-versioning).

#### _Global Workflow Versioning_
Some users do "Global Versioning" with Workflow definitions wherein only new Executions receive the new behavior
you must Version ([source](https://medium.com/@qlong/how-to-overcome-some-maintenance-challenges-of-temporal-cadence-workflow-versioning-f893815dd18d)).
This works if you do not have long-running Workflows that need to pick up changes after a blocking Awaitable(eg, `Timer` usage). 
For example, you might have a "Subscription" Workflow that sleeps for three months and performs an Activity upon `TimerFired`.
If you want to add an Activity post-timer, you likely want to pick up this new Activity - even in Workflows that have been

> `Open` for some time. This likely means falling back to using `GetVersion` directly or using **ContinueAsNew** to force a new

Reviewer: ```suggestion
open for some time. This likely means falling back to using `GetVersion` directly or using `ContinueAsNew` to force a new
```

execution with input parameters that reflect what work has already been completed.

## Workflow Versioning: Routed Strategy

You can target specific Versions of your Workflow Definition from the `Starter` by either:

> 1. Specifying the WorkflowType explicitly; eg `MyWorkflowV2`

Reviewer: ```suggestion
1. Specifying the WorkflowType explicitly; e.g., `MyWorkflowV2`
```

> 2. Designating a TaskQueue for where the new Workflow Version is hosted; eg `TaskQueue: "PaymentsV2"`

Reviewer: ```suggestion
2. Designating a Task Queue for where the new Workflow Version is hosted; e.g., `TaskQueue: "PaymentsV2"`
```

> Both strategies will avoid the risk of **NonDeterminismExceptions** in currently Open Workflows.

Reviewer: ```suggestion
Both strategies will avoid the risk of a `NonDeterministicException` in currently Open Workflows.
```

There are a few subtle differences to consider to decide which is suitable.

#### WorkflowType Routed Versioning

> This type of Versioning appends a Version into the Implementation name; eg `MyWorkflowV2`.

Reviewer: ```suggestion
This type of Versioning appends a Version into the Implementation name; e.g., `MyWorkflowV2`.
```

This is by convention only and is not enforced.

_Pros_

* Does not demand changes to your Worker deployment.
* Allows for discrete changes to each Workflow Type within the Application.

_Cons_

> * Starters must update `Start/Execute` Workflow code to route to the new Implementation, possibly forcing cross-team deployments.

Reviewer: ```suggestion
* Starters must update `Start/Execute` Workflow code to route to the new implementation, possibly forcing cross-team deployments.
```

> * Requires custom **SearchAttribute** or metrics to determine when the old Implementations can be deregistered.

Reviewer: ```suggestion
* Requires custom **SearchAttribute** or metrics to determine when the old implementations can be deregistered.
```

* Changing definition behavior results in a whole new source file, making Git diffs exposing the changes more difficult to spot in PRs.

##### Removing Old WorkflowType Versions

Simply remove the registration of the WorkflowType from relevant Workers.

#### TaskQueue Routed Versioning

_Pros_

* Might be easier to determine which Version of your application to decommission by observing Worker compute resources.
* Explicit Version support at a higher Application level since multiple Workflow Types can be registered at once.

_Cons_

> * Starters must update `Start/Execute` Workflow code to route to the new Implementation, possibly forcing cross-team deployments.

Reviewer: ```suggestion
* Starters must update `Start/Execute` Workflow code to route to the new implementation, possibly forcing cross-team deployments.
```

* Worker count will increase exponentially, increasing complexity in infrastructure deployments and observability.
  * This is further exacerbated if using `TaskQueue` for multi-tenancy purposes.

##### Removing Old TaskQueue Versions

If you have multiple WorkerFactories (Java) in a process, remove these factories
using the deprecated TaskQueue. If a single WorkerFactory is in a process, remove it from
the deployment.

## Replay Tests

Recall that Workflow Executions are "replayed" at various times during your Application lifetime.

> This means we must verify that the state and behaviors meet expectations regardless of **how** it is executed.

Your Workflow could build, pass all unit and functional tests, and ship to production just fine, 
but still bring your business to a halt because of Non-Determinism Errors.

> This incident will be now exacerbated by the _new_ Workflow Executions that have started in production while the previous versions were "stuck".

Reviewer: ```suggestion
This incident will be now exacerbated by the _new_ Workflow Executions that have started in production while the previous versions were "stuck."
```

executions in your development or build pipeline.

These types of tests are a variant of "smoke" or "build validation" tests, so run these tests where you are currently performing these other tests; 
likely this is a step just before running your other unit or functional tests.
Alternately, you can run these tests as a *verification* step and might put these in the developer code commit workflow as early as a `git commit`.

Regardless of your environment or where in the delivery process these appear, they are an important investment to your production quality.

#### Recommendations

Replay tests should exercise history that was produced by various code paths through your Executions.
If your Workflow Definition has conditional branches, loops, or timers, it makes sense to store histories created by the unit test cases that verify these scenarios with the
execution histories _caused_ by the scenarios in the `latest` build so that they might be validated in `vNext` implementations.

Factors which determine how these tests are executed are based on your Workflow Version strategy and isolation boundaries present in your

> Continuous Integration / Continuous Deployment (CICD) pipeline.

Reviewer: ```suggestion
Continuous Integration / Continuous Deployment (CI/CD) pipeline.
```

GitHub, GitLab, and RedHat all include the slash in the abbreviation.

#### Implementing Replay Tests : Patched Version Strategy

If you elect to use the `GetVersion` or `patch` SDK APIs, you must choose between loading Workflow History representing the
`latest` build from memory or from file storage.

**External Workflow History Storage**

* _Cons_

>     * This can be problematic in ephemeral containers used for CICD purposes.

Reviewer: ```suggestion
    * This can be problematic in ephemeral containers used for CI/CD purposes.
```

    * This might require non-trivial authnz to an external file storage solution.
    * Garbage collection on histories might be needed to keep storage costs down

>         * Detecting which histories are eligible for collection can be complex for Workflow Definitions that are long-running and do not implement Continue As New

Reviewer: ```suggestion
        * Detecting which histories are eligible for collection can be complex for Workflow Definitions that are long-running and do not implement Continue-As-New
```

    * The `DataConverter` used to generate the histories _must_ be used when doing these tests. 
* _Pros_

>     * No need to maintain more than one implementation in VCS for the same Workflow Type in the same source version.

Reviewer: ```suggestion
    * No need to maintain more than one implementation in a version control system (VCS) for the same Workflow Type in the same source version.
```

I suspect that some developers might not be familiar with the abbreviation, so I recommend spelling it out on first reference here. In my experience, some (older) people say "revision control system." I suspect that some younger people have only ever known git and therefore have no need to generalize about what kind of tool it is.

        * (see note on Worker registration considerations below)
    * Duplicate implementations of same representation in same VCS version might philosophically contradict your organization VCS principles.
        * Placing a version in the Workflow Type name "hides" implementation changes in git comparisons by breaking helpful commit diffs during code review.
        * The git concern can be mitigated with a descending versioning scheme, where the `vNext` overwrites the `latest` while preserving the same filename.
    * Long-running workflows that do not use ContinueAsNew will keep their history around for a while so quite old implementations will need to survive in VCS to be validated against the proposed `vNext` implementation.
* _Pros_
    * No need to maintain external storage solution for the history produced by `latest` build. Validation can be done via in-memory history produced in unit tests.

> The Patched Versioning strategy is convenient because `Starters` don't need to update their code with version changes or coordinate deployments with services hosting Temporal Workers.
>

> > Note that in all SDKs except Java need to specify a `string` WorkflowType name explicitly in the Worker registration,type decorator or attribute options

Reviewer: ```suggestion
> Note that in all SDKs except Java need to specify a `string` Workflow Type name explicitly in the Worker registration,type decorator or attribute options
```
